src/core/server.cc

```diff
@@ -472,58 +675,28 @@ HTTPServiceImpl::Infer(
         infer_request_header_str, &request_provider);
     if (status.ok()) {
       infer_stats->SetBatchSize(request_provider->RequestHeader().batch_size());
 
       std::shared_ptr<HTTPInferResponseProvider> response_provider;
       status = HTTPInferResponseProvider::Create(
-          req->OutputBuffer(), *(backend->Backend()),
+          req->buffer_out, *(backend->Backend()),
           request_provider->RequestHeader(), &response_provider);
       if (status.ok()) {
+        evhtp_connection_t* htpconn = evhtp_request_get_connection(req);
+        evthr_t* thread = htpconn->thread;
+        ThreadWiseService* requests = (ThreadWiseService*)evthr_get_aux(thread);
+        ThreadWiseService::Request* request =
+            requests->Enqueue(req, id, request_provider, response_provider);
         server_->HandleInfer(
-            &request_status, backend, request_provider, response_provider,
-            infer_stats,
-            [&request_status, id, request_provider, response_provider,
-             infer_stats, req]() mutable {
-              if (request_status.code() == RequestStatusCode::SUCCESS) {
-                std::string format;
-                if (!req->QueryParam("format", &format)) {
-                  format = "text";
-                }
-
-                // The description of the raw outputs needs to go in
-                // the kInferResponseHTTPHeader since it is needed to
-                // interpret the body. The entire response (including
-                // classifications) is serialized at the end of the
-                // body.
-                InferResponseHeader* response_header =
-                    response_provider->MutableResponseHeader();
-                response_header->set_id(id);
-
-                std::string rstr;
-                if (format == "binary") {
-                  response_header->SerializeToString(&rstr);
-                } else {
-                  rstr = response_header->DebugString();
-                }
-                req->WriteResponseBytes(rstr.c_str(), rstr.size());
-
-                // We do this in destructive manner since we are the
-                // last one to use response header from the provider.
-                for (int i = 0; i < response_header->output_size(); ++i) {
-                  InferResponseHeader::Output* output =
-                      response_header->mutable_output(i);
-                  output->clear_batch_classes();
-                }
-
-                req->OverwriteResponseHeader(
-                    kInferResponseHTTPHeader,
-                    response_header->ShortDebugString());
-              }
+            &(request->request_status), backend, request->request_provider,
+            request->response_provider, infer_stats,
+            [requests, request]() mutable {
+              requests->FinalizeHTTPResponse(request);
             },
-            false  // async frontend
+            true  // async frontend
```

Comment:
Not in this pull request, but in a follow-up pull request should remove this option and clean up the downstream synchronous path.